sycl/plugins/level_zero/pi_level_zero.hpp

```diff
@@ -826,13 +844,24 @@ struct _pi_program : _pi_object {
 
 struct _pi_kernel : _pi_object {
   _pi_kernel(ze_kernel_handle_t Kernel, pi_program Program)
-      : ZeKernel{Kernel}, Program{Program} {}
+      : ZeKernel{Kernel}, Program{Program}, MemAllocs{}, UsersCount{0} {}
 
   // Level Zero function handle.
   ze_kernel_handle_t ZeKernel;
 
   // Keep the program of the kernel.
   pi_program Program;
+
+  // If kernel has indirect access we need to make a snapshot of all existing
+  // memory allocations to defer deletion of these memory allocations to the
+  // moment when kernel execution has finished.
+  std::list<MemAllocRecord *> MemAllocs;
```

Comment:
Why do we need to take the snapshot instead of just reference-counting the allocation? I.e. you'd retain all active allocations at kernel enqueue, and release at kernel completion.

Reply:
Because we need to know which memory allocations to release at kernel completion. Picture of active allocations can change during kernel execution.

Reply:
I think I understand now, but please clarify it in source comment.

Reply:
Fixed.